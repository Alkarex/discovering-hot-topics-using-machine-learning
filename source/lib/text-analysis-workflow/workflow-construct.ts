/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICNSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/


import { Construct, Aws } from '@aws-cdk/core';
import { StateMachine, Chain, LogLevel, CfnStateMachine, StateMachineType } from '@aws-cdk/aws-stepfunctions';
import { CfnPolicy, Effect, Policy, PolicyStatement, Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { LogGroup, LogStream } from '@aws-cdk/aws-logs';
import { DefaultLogGroupProps } from '@aws-solutions-constructs/core';
export interface WorkflowProps {
    stateMachineType?: StateMachineType;
    chain: Chain
}
export class Workflow extends Construct {
    private _stMachine: StateMachine

    constructor (scope: Construct, id: string, props: WorkflowProps) {
        super(scope, id);
        
        let _stateMachineRole: Role;

        if (props.stateMachineType === StateMachineType.EXPRESS) {
            const logGroup: LogGroup = new LogGroup(this, 'statemachine-log-group', DefaultLogGroupProps());
            const logStream: LogStream = logGroup.addStream('statemachine-log-stream');

            _stateMachineRole = new Role(this, 'StateMachineRole', {
                assumedBy: new ServicePrincipal(`states.${Aws.REGION}.amazonaws.com`)
            });

            const stateMachineLogPolicy = new Policy(this, 'StateMachineLogPolicy');
            stateMachineLogPolicy.addStatements(new PolicyStatement({
                actions: [                
                    "logs:CreateLogDelivery",
                    "logs:GetLogDelivery",
                    "logs:UpdateLogDelivery",
                    "logs:DeleteLogDelivery",
                    "logs:ListLogDeliveries",
                    "logs:PutResourcePolicy",
                    "logs:DescribeResourcePolicies",
                    "logs:DescribeLogGroups"
                ],
                resources: [ '*' ],
                effect: Effect.ALLOW
            }));

            stateMachineLogPolicy.attachToRole(_stateMachineRole);
            (stateMachineLogPolicy.node.defaultChild as CfnPolicy).cfnOptions.metadata = {
                cfn_nag: {
                    rules_to_suppress: [{
                        id: 'W12',
                        reason: 'The stepfunction log policy requires that resources be "*"'
                    }]
                }
            };

            this._stMachine = new StateMachine(this, 'WorkflowEngine', {
                definition: props.chain,
                ...(props.stateMachineType === StateMachineType.EXPRESS && {
                    stateMachineType: props.stateMachineType,
                    role: _stateMachineRole
                })
            });

            const cfnStateMachine = this._stMachine.node.defaultChild as CfnStateMachine;
            cfnStateMachine.addPropertyOverride(
                'LoggingConfiguration', {
                    Destinations: [{
                        CloudWatchLogsLogGroup: {
                            LogGroupArn: logGroup.logGroupArn
                        }
                    }],
                    IncludeExecutionData: true,
                    Level: LogLevel.ALL
                }
            );

        } else {
            this._stMachine = new StateMachine(this, 'WorkflowEngine', {
                definition: props.chain
            });
        }
    }

    public get stateMachine(): StateMachine {
        return this._stMachine;
    }
}