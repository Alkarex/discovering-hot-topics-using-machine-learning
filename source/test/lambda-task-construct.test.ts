/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICNSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import { StepFuncLambdaTask } from '../lib/text-analysis-workflow/lambda-task-construct';

import '@aws-cdk/assert/jest';

test('test lambda task construct', () => {
    const stack = new cdk.Stack();

    new  StepFuncLambdaTask (stack, 'StepFuncWorkflowStack', {
        taskName: 'unitTestTask',
        lambdaFunctionProps: {
            runtime: lambda.Runtime.NODEJS_12_X,
            handler: 'index.handler',
            code: lambda.Code.asset(`${__dirname}/../lambda/wf-analyze-text`)
        },
    });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});