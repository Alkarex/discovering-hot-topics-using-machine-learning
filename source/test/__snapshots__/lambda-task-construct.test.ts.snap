// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test lambda task construct 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters399d769f91488076c514f62b071c692fe78c8108ee208544cd5315d18bc48cfdArtifactHash1B323516": Object {
      "Description": "Artifact hash for asset \\"399d769f91488076c514f62b071c692fe78c8108ee208544cd5315d18bc48cfd\\"",
      "Type": "String",
    },
    "AssetParameters399d769f91488076c514f62b071c692fe78c8108ee208544cd5315d18bc48cfdS3Bucket28277303": Object {
      "Description": "S3 bucket for asset \\"399d769f91488076c514f62b071c692fe78c8108ee208544cd5315d18bc48cfd\\"",
      "Type": "String",
    },
    "AssetParameters399d769f91488076c514f62b071c692fe78c8108ee208544cd5315d18bc48cfdS3VersionKeyE094BFBE": Object {
      "Description": "S3 key for asset version \\"399d769f91488076c514f62b071c692fe78c8108ee208544cd5315d18bc48cfd\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "StepFuncWorkflowStackLambdaFunctionEA7C37F1": Object {
      "DependsOn": Array [
        "StepFuncWorkflowStackLambdaFunctionServiceRoleDefaultPolicyC18D2E91",
        "StepFuncWorkflowStackLambdaFunctionServiceRoleD7BAFDA7",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W58",
              "reason": "Lambda functions has the required permission to write CloudWatch Logs. It uses custom policy instead of arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole with more tighter permissions.",
            },
          ],
        },
      },
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters399d769f91488076c514f62b071c692fe78c8108ee208544cd5315d18bc48cfdS3Bucket28277303",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters399d769f91488076c514f62b071c692fe78c8108ee208544cd5315d18bc48cfdS3VersionKeyE094BFBE",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters399d769f91488076c514f62b071c692fe78c8108ee208544cd5315d18bc48cfdS3VersionKeyE094BFBE",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "StepFuncWorkflowStackLambdaFunctionServiceRoleD7BAFDA7",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "StepFuncWorkflowStackLambdaFunctionServiceRoleD7BAFDA7": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:aws:logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "LambdaFunctionServiceRolePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "StepFuncWorkflowStackLambdaFunctionServiceRoleDefaultPolicyC18D2E91": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "Lambda needs the following minimum required permissions to send trace data to X-Ray.",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StepFuncWorkflowStackLambdaFunctionServiceRoleDefaultPolicyC18D2E91",
        "Roles": Array [
          Object {
            "Ref": "StepFuncWorkflowStackLambdaFunctionServiceRoleD7BAFDA7",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
